doc/Overview.md

```diff
@@ -0,0 +1,81 @@
+Overview
+======
+
+Note: This describes the layout of the Tock 2.0 `libtock-rs` crates, not the
+Tock 1.0 crates. Not all of the Tock 2.0 crates have been merged into this
+repository yet.
+
+This document gives an overview of the crates in this repository, and is
+intended to be useful to `libtock-rs` newcomers.
+
+## `libtock`
+
+Note: `libtock` is currently called `libtock2`, as `libtock` is already a Tock
+1.0 crate.
+
+`libtock` provides the default `libtock-rs` experience. It re-exports all of the
+drivers `libtock-rs` provides, and provides usable defaults for panic handling
+and memory allocation. It should be easy to build a Tock application that only
+has one direct dependency: `libtock`.
+
+In order to be easy to use, `libtock` lacks some functionality that the other
+`libtock-rs` crates provide. In particular, you should depend on the other
+crates directly if you want any of the following:
+
+* The ability to unit test your code.
+* The ability to customize panic handling.
+* The ability to run without dynamic memory allocation or with an alternative
```

Comment:
This is a bit of a loss relative to the current `libtock`, which (I think?) allows apps to forego dynamic memory allocation, but this generally seems fine -- anyone highly worried about dependable/efficient apps is unlikely to use `libtock`

Reply:
Hmm, this was actually the result of a misunderstanding on my part, and I was performing an optimization prematurely.

I noticed that running [`make examples`](https://github.com/tock/libtock-rs/blob/84c4bf25b784f26b9cac86faad9247a61ba6244b/Makefile#L79) after making a minor change to `libtock-rs` takes a while. This made `make test` slow, and was impeding my development. I *assumed* the issue was that changing feature flags invalidates `cargo`'s build artifacts, causing complete rebuilds. As a result, I was trying to minimize the use of feature flags in `libtock-rs`.

I decided verify my assumption before responding to your comment -- and my assumption was wrong. `cargo` caches build artifacts with different feature combinations perfectly fine.

I'll remove that line; `libtock` can support dynamic memory allocation behind a feature flag.

Reply:
I actually removed both the line about dynamic memory allocation and switchable panic handlers: we can expose both sets of functionality behind a cargo feature flag.

I'm hopeful that because `libtock-rs` will consist of a larger number of small crates (most of which do not have feature flags), cargo's cache will be more effective in CI. I'm not 100% sure of this, though, because most of the driver code will be monomorphized in `libtock`.

Reply:
Awesome, thanks!

Reply:
I did some more testing, and there's a hidden footgun with `cargo` features.

I set up 3 libraries:

1. `libtock`, which has a `malloc` feature.
2. `needs_malloc`, which depends on `libtock` and needs the `malloc` feature.
3. `no_malloc`, which depends on `libtock` but does not need the `malloc` feature.

`no_malloc` and `needs_malloc` are in the same workspace.

As expected, `needs_malloc` always has access to `malloc`. However, if you use `cargo build -p no_malloc -p needs_malloc` to build both binaries, `no_malloc` gets the `malloc` feature as well!

I'm not sure this footgun is a deal-breaker, but it at least needs to be considered and documented if we choose to make malloc optional.

Reply:
I am guessing this means that building the entire workspace (e.g. by running `cargo build` from the top level) will lead to some incorrectly built binaries? This does seem a little messy, and we should definitely document it well.

Reply:
Yes, assuming that LTO is unable to remove the memory allocation implementation.

I was planning to build a hook into `libtock_runtime` so it can initialize a memory allocator before `main` begins. However, if you do that, LTO will not be able to remove the memory allocation initialization. An alternative is to make our allocators initialize themselves when they're first used, which may add a bit of overhead in some cases.

I'm leaning towards documenting it. I think that should be a separate document in a separate PR.

Reply:
documenting sounds better than adding artificial overhead, and a separate PR sounds fine